pkg/dataobj/index/builder.go

```diff
@@ -251,8 +251,12 @@ func (p *Builder) run(ctx context.Context) error {
 	level.Info(p.logger).Log("msg", "started index builder service")
 	for {
 		fetches := p.client.PollRecords(ctx, -1)
-		if fetches.IsClientClosed() || ctx.Err() != nil {
-			return ctx.Err()
```

Comment:
When `fetches.IsClientClosed()` is true, it's often also the case that `ctx.Err()` is `nil`, as the context won't have been canceled. This fixes returning a nil error for closed clients.